File: 2402-meeting-rooms-iii/2402-meeting-rooms-iii.java
class Solution {
    public int mostBooked(int n, int[][] meetings) {
        int meetingCount[] = new int[n];
        PriorityQueue<Integer> unused = new PriorityQueue<>();
        PriorityQueue<long[]> used = new PriorityQueue<>(
            (a, b) -> a[0] != b[0] ? Long.compare(a[0], b[0]) : Long.compare(a[1], b[1])
        );
        for(int i = 0;i < n;i++){
            unused.offer(i);
        }
        Arrays.sort(meetings, (a, b) -> a[0] != b[0] ? Integer.compare(a[0], b[0]) : Integer.compare(a[1], b[1]));
        for(int meeting[] : meetings){
            int start = meeting[0];
            int end = meeting[1];
            while(!used.isEmpty() && used.peek()[0] <= start){
                int room = (int)used.poll()[1];
                unused.offer(room);
            }
            if(!unused.isEmpty()){
                int room = unused.poll();
                used.offer(new long[]{end, room});
                meetingCount[room]++;
            }else{
                long roomAvalTime = used.peek()[0];
                int room = (int)used.poll()[1];
                used.offer(new long[]{roomAvalTime + end - start, room});
                meetingCount[room]++;
            }
        }
        int maxCount = 0;
        int maxRoom = 0;
        for(int i = 0;i < n;i++){
            if(meetingCount[i] > maxCount){
                maxCount = meetingCount[i];
                maxRoom = i;
            }
        }
        return maxRoom;
    }
}

File: 2402-meeting-rooms-iii/2402-meeting-rooms-iii.py
class Solution:
    def mostBooked(self, n: int, meetings: List[List[int]]) -> int:
         
        meetings.sort()
        hashmap = defaultdict(int)
        heap = [[0, i] for i in range(n)]
        heapify(heap)

        for start, end in meetings:
            for i in range(n):
                if heap[i][0] <= start:
                    heap[i][0] = 0
            heapify(heap)
            if heap[0][0] <= start:
                hashmap[heap[0][1]] += 1
                heappushpop(heap, [end, heap[0][1]])
            else:
                hashmap[heap[0][1]] += 1
                heappushpop(heap, [end - start + heap[0][0], heap[0][1]])

        res = hashmap[0]
        output = 0
        for key in hashmap:
            if res < hashmap[key]:
                output = key
                res = hashmap[key]
            elif res == hashmap[key]:
                output = min(key, output)
        
        return output